core/src/main/java/io/ddf/Factor.java

```diff
@@ -171,38 +206,22 @@ public void setLevels(List<String> levels) throws DDFException {
    *                  meaning
    * @throws DDFException
    */
-  public void setLevels(List<String> levels, List<Integer> codes, boolean isOrdered) throws DDFException {
+  public void setLevels(List<Object> levels, List<Integer> codes, boolean isOrdered) throws DDFException {
     if (levels == null || levels.isEmpty()) throw new DDFException("Levels cannot be null or empty");
     if (codes != null && codes.size() != levels.size()) throw new DDFException(String.format(
         "The number of levels is %d which does not match the number of codes %d", levels.size(), codes.size()));
-
-    if (mLevelMap == null) mLevelMap = this.instantiateSynchronizedLevelMap();
-
-    if (codes == null) {
-      // Auto-create a 1-based level-code map
-      codes = new ArrayList<Integer>();
-      for (int i = 1; i <= levels.size(); i++) {
-        codes.add(i);
+    if(codes != null) {
+      if (mLevelMap == null) mLevelMap = this.instantiateSynchronizedLevelMap();
```

Comment:
Even with a synchronized map, this Factor class won't be thread safe. It's mutable, keep it's state in multiple fields and there is no synchronization on all the methods that mutate its state.